qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,657 @@
+
+        # Dict of the form: (schedule.name, parameter.name, qubits): Parameter
+        self._parameter_map = {}
+
+        # Reverse mapping of _parameter_map
+        self._parameter_map_r = defaultdict(set)
+
+        # Default dict of the form: (schedule.name, parameter.name, qubits): [ParameterValue, ...]
+        self._params = defaultdict(list)
+
+        self._schedules = {}
+
+    def add_schedule(self, schedule: Schedule, qubits: Tuple = None):
+        """
+        Add a schedule and register its parameters.
+
+        Args:
+            schedule: The schedule to add.
+            qubits: The qubits for which to add the schedules. If None is given then this
+                schedule is the default schedule for all qubits.
+
+        Raises:
+            CalibrationError: If the parameterized channel index is not formatted
+                following index1.index2... or if several parameters in the same schedule
+                have the same name.
+        """
+        # check that channels, if parameterized, have the proper name format.
+        # pylint: disable = raise-missing-from
+        param_indices = set()
+        for ch in schedule.channels:
+            if isinstance(ch.index, Parameter):
+                param_indices.add(ch.index)
+                if re.compile(self.__channel_pattern__).match(ch.index.name) is None:
+                    raise CalibrationError(
+                        f"Parameterized channel must correspond to {self.__channel_pattern__}"
+                    )
+
+        self._schedules[(schedule.name, qubits)] = schedule
+
+        param_names = [param.name for param in schedule.parameters]
+
+        if len(param_names) != len(set(param_names)):
+            raise CalibrationError(f"Parameter names in {schedule.name} must be unique.")
+
+        # Register parameters that are not indices.
+        for param in schedule.parameters:
+            if param not in param_indices:
+                self._register_parameter(param, schedule, qubits)
+
+    def _register_parameter(
+        self, parameter: Parameter, schedule: Schedule = None, qubits: Tuple = None
+    ):
+        """
+        Registers a parameter for the given schedule. This allows self to determine the
+        parameter instance that corresponds to the given schedule name, parameter name
+        and qubits.
+
+        Args:
+            parameter: The parameter to register.
+            schedule: The Schedule to which this parameter belongs. The schedule can
+                be None which allows the calibration to accommodate, e.g. qubit frequencies.
+            qubits: The qubits for which to register the parameter.
+        """
+        sched_name = schedule.name if schedule else None
+        key = ParameterKey(sched_name, parameter.name, qubits)
+        self._parameter_map[key] = parameter
+        self._parameter_map_r[parameter].add(key)
+
+    @property
+    def parameters(self) -> Dict[Parameter, Set[ParameterKey]]:
```

Comment:
This is used in the `to_csv`. It could be made private but it is nice to have public so that users can understand/debug their calibrations if need be.

Reply:
I think it is fine for it to be public. I just wondered if the name `parameters` might be better used on a different quantity. I was trying to think of what the most useful thing to the user would be related to "parameters." The best I came up with was something like the output of `parameter_table(None, None, None)` but limited to the most recent valid value for each entry.